### PR TITLE
Speed up summary_config by converting ecl_sum to datetime in C++

### PR DESCRIPTION
**Pull request by JHolba**
based on #6217. 
updated to use manual conversion to python datetime instead of pybind11 doing it automatically.

## Pre review checklist

- [x] Read through the code changes carefully after finishing work
- [x] Make sure tests pass locally (after every commit!)
- [x] Prepare changes in small commits for more convenient review (optional)
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Updated documentation
- [x] Ensured that unit tests are added for all new behavior (See 
    [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)),
    and changes to existing code have good test coverage.

## Pre merge checklist
- [ ] Added appropriate release note label
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: src/clib/lib/enkf/read_summary.cpp [hunk 1]
```diff
@@ -16,7 +20,27 @@ static bool matches(std::vector<std::string> patterns, std::string key) {
...
+
+        time_t_vector_type *tvec = ecl_sum_alloc_time_vector(summary, true);
+        int size = time_t_vector_size(tvec);
+        std::vector<pybind11::object> result(size);
+        auto t = tm{};
+        for (int i = 0; i < size; i++) {
+            auto i_time = time_t_vector_iget(tvec, i);
+            ::gmtime_r(&i_time, &t);
+            result[i] = pybind11::reinterpret_steal<pybind11::object>(
+                PyDateTime_FromDateAndTime(t.tm_year + 1900, t.tm_mon + 1,
```
Comment (eivindjahren) [inline src/clib/lib/enkf/read_summary.cpp:38]: the cpp linter is angry

Comment (JHolba) [reply src/clib/lib/enkf/read_summary.cpp]: the cpp linter didn't understand the PyDateTime_IMPORT I believe. Added an if in front to placate the linter.

File: src/clib/lib/enkf/read_summary.cpp [hunk 2]
```diff
@@ -1,25 +1,58 @@
...
 }
+
 ERT_CLIB_SUBMODULE("_read_summary", m) {
+    m.def("read_dates", [](Cwrap<ecl_sum_type> summary) {
+        if (!PyDateTimeAPI)
+            PyDateTime_IMPORT;
+
+        time_t_vector_type *tvec = ecl_sum_alloc_time_vector(summary, true);
+        int size = time_t_vector_size(tvec);
+        std::vector<pybind11::object> result(size);
```
Comment (pinkwah) [inline src/clib/lib/enkf/read_summary.cpp:29]: Maybe use a `py::list` directly? No need to use an `std::vector` if it ends up as a Python `list` in the end anyway.

File: src/clib/lib/enkf/read_summary.cpp [hunk 3]
```diff
@@ -1,25 +1,58 @@
...
 static bool matches(std::vector<std::string> patterns, std::string key) {
-    bool has_key = false;
-    for (auto pattern : patterns) {
-        if (fnmatch(pattern.c_str(), key.c_str(), 0) == 0) {
-            has_key = true;
-            break;
-        }
-    }
-    return has_key;
+    return std::any_of(patterns.cbegin(), patterns.cend(),
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/clib/lib/enkf/read_summary.cpp:16]: I would prefer `std::any_of` tbf.

File: src/clib/lib/enkf/read_summary.cpp [hunk 4]
```diff
@@ -1,25 +1,58 @@
...
+                           return fnmatch(pattern.c_str(), key, 0) == 0;
+                       });
 }
+
 ERT_CLIB_SUBMODULE("_read_summary", m) {
+    m.def("read_dates", [](Cwrap<ecl_sum_type> summary) {
+        if (!PyDateTimeAPI)
+            PyDateTime_IMPORT;
+
+        time_t_vector_type *tvec = ecl_sum_alloc_time_vector(summary, true);
```
Comment (eivindjahren) [inline src/clib/lib/enkf/read_summary.cpp:27]: You have to do `time_t_vector_free` on this.

**Review by eivindjahren — APPROVED** [review]

**Pull request merged**